File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.py
import heapq
from collections import defaultdict
from typing import List

class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start_node: int, end_node: int) -> float:
        # Step 1: Create an adjacency list for the graph
        graph = defaultdict(list)
        for i, (a, b) in enumerate(edges):
            prob = succProb[i]
            graph[a].append((b, prob))
            graph[b].append((a, prob))
        
        # Step 2: Initialize max heap and max probability array
        max_heap = [(-1.0, start_node)]  # We use negative to simulate max heap
        max_prob = [0.0] * n
        max_prob[start_node] = 1.0
        
        # Step 3: Dijkstra's algorithm using max heap
        while max_heap:
            curr_prob, node = heapq.heappop(max_heap)
            curr_prob = -curr_prob  # Convert back to positive
            
            # If we reach the end node, return the probability
            if node == end_node:
                return curr_prob
            
            # Explore neighbors
            for neighbor, prob in graph[node]:
                new_prob = curr_prob * prob
                if new_prob > max_prob[neighbor]:
                    max_prob[neighbor] = new_prob
                    heapq.heappush(max_heap, (-new_prob, neighbor))
        
        return 0.0

File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.java
import java.util.*;

class Solution {
    public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
        // Create an adjacency list
        Map<Integer, List<Pair>> graph = new HashMap<>();
        for (int i = 0; i < n; i++) {
            graph.put(i, new ArrayList<>());
        }
        
        for (int i = 0; i < edges.length; i++) {
            int a = edges[i][0];
            int b = edges[i][1];
            double prob = succProb[i];
            graph.get(a).add(new Pair(b, prob));
            graph.get(b).add(new Pair(a, prob));
        }
        
        // Priority queue to simulate a max-heap (negate the probabilities)
        PriorityQueue<Pair> maxHeap = new PriorityQueue<>((a, b) -> Double.compare(b.prob, a.prob));
        maxHeap.offer(new Pair(start_node, 1.0));
        
        // Track maximum probabilities
        double[] maxProb = new double[n];
        maxProb[start_node] = 1.0;
        
        // Process the queue
        while (!maxHeap.isEmpty()) {
            Pair current = maxHeap.poll();
            int node = current.node;
            double prob = current.prob;
            
            if (node == end_node) {
                return prob;
            }
            
            for (Pair neighbor : graph.get(node)) {
                int nextNode = neighbor.node;
                double edgeProb = neighbor.prob;
                double newProb = prob * edgeProb;
                
                if (newProb > maxProb[nextNode]) {
                    maxProb[nextNode] = newProb;
                    maxHeap.offer(new Pair(nextNode, newProb));
                }
            }
        }
        
        return 0.0;
    }
    
    // Helper class to store pairs of (node, probability)
    private static class Pair {
        int node;
        double prob;
        
        Pair(int node, double prob) {
            this.node = node;
            this.prob = prob;
        }
    }
}
